pkg/endpoint/redirect_test.go

```diff
@@ -0,0 +1,256 @@
+type DummyIdentityAllocatorOwner struct{}
+
+// UpdateIdentities does nothing.
+func (d *DummyIdentityAllocatorOwner) UpdateIdentities(added, deleted cache.IdentityCache) {
+	return
+}
+
+// GetNodeSuffix does nothing.
+func (d *DummyIdentityAllocatorOwner) GetNodeSuffix() string {
+	return ""
+}
+
+// DummyOwner implements pkg/endpoint/regeneration/Owner. Used for unit testing.
+type DummyOwner struct {
+	repo *policy.Repository
+}
+
+// GetPolicyRepository returns the policy repository of the owner.
+func (d *DummyOwner) GetPolicyRepository() *policy.Repository {
+	return d.repo
+}
+
+// QueueEndpointBuild does nothing.
+func (d *DummyOwner) QueueEndpointBuild(ctx context.Context, epID uint64) (func(), error) {
+	return nil, nil
+}
+
+// GetCompilationLock does nothing.
+func (d *DummyOwner) GetCompilationLock() *lock.RWMutex {
+	return nil
+}
+
+// SendNotification does nothing.
+func (d *DummyOwner) SendNotification(typ monitorAPI.AgentNotification, text string) error {
+	return nil
+}
+
+// Datapath returns a nil datapath.
+func (d *DummyOwner) Datapath() datapath.Datapath {
+	return nil
+}
+
+// GetNodeSuffix does nothing.
+func (d *DummyOwner) GetNodeSuffix() string {
+	return ""
+}
+
+// UpdateIdentities does nothing.
+func (d *DummyOwner) UpdateIdentities(added, deleted cache.IdentityCache) {}
+
+func (s *RedirectSuite) TestAddVisibilityRedirects(c *check.C) {
+	// Setup dependencies for endpoint.
+	kvstore.SetupDummy("etcd")
+	defer kvstore.Close()
+
+	identity.InitWellKnownIdentities()
+	idAllocatorOwner := &DummyIdentityAllocatorOwner{}
+
+	<-cache.InitIdentityAllocator(idAllocatorOwner, nil, nil)
+	defer cache.Close()
+
+	do := &DummyOwner{
+		repo: policy.NewPolicyRepository(),
+	}
+	identitymanager.Subscribe(do.repo)
+
+	lblQA := labels.ParseLabel("QA")
+	lblBar := labels.ParseLabel("bar")
+
+	httpPort := uint16(19001)
+	dnsPort := uint16(19002)
+	kafkaPort := uint16(19003)
+
+	rsp := &RedirectSuiteProxy{
+		parserProxyPortMap: map[policy.L7ParserType]uint16{
+			policy.ParserTypeHTTP:  httpPort,
+			policy.ParserTypeDNS:   dnsPort,
+			policy.ParserTypeKafka: kafkaPort,
+		},
+		redirectPortUserMap: make(map[uint16][]string),
+	}
+
+	ep := NewEndpointWithState(do, rsp, 12345, StateRegenerating)
+
+	qaBarLbls := labels.Labels{lblBar.Key: lblBar, lblQA.Key: lblQA}
+	epIdentity, _, err := cache.AllocateIdentity(context.Background(), idAllocatorOwner, qaBarLbls)
+	c.Assert(err, check.IsNil)
+	ep.SetIdentity(epIdentity, true)
+
+	firstAnno := map[string]string{
+		annotation.ProxyVisibility: "<Ingress/80/TCP/HTTP>",
+	}
+	ep.UpdateVisibilityPolicy(firstAnno)
+	err = ep.regeneratePolicy()
+	c.Assert(err, check.IsNil)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	cmp := completion.NewWaitGroup(ctx)
+
+	_, err, _, _ = ep.addNewRedirects(ep.desiredPolicy.L4Policy, cmp)
+	c.Assert(err, check.IsNil)
+	v, ok := ep.desiredPolicy.PolicyMapState[policy.Key{
+		Identity:         0,
```

Comment:
Why is identity zero here?

Reply:
Identity of zero is an allow-all in our datapath. Proxy redirection occurs for a given L4-port / protocol - this means it allows-all at L3.

Reply:
We should probably add a constant for this somewhere ;)